Utils/WeakList.swift

```diff
@@ -0,0 +1,66 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+/**
+ * A list that weakly holds references to its items.
+ * Note that while the references themselves are cleared, their wrapper objects
+ * are not (though they are reused). Also note that since slots are reused,
+ * order is not preserved.
+ *
+ * This class crashes at runtime with EXC_BAD_ACCESS if a protocol is given as
+ * the type T. Make sure to use a class type.
+ */
+public class WeakList<T: AnyObject>: SequenceType {
+    private var items = [WeakRef<T>]()
+
+    public init() {}
+
+    /**
+     * Adds an item to the list.
+     * Note that every insertion iterates through the list to find any "holes" (items that have
+     * been deallocated) to reuse them, so this class may not be appropriate in situations where
+     * appending is frequent.
+     */
+    public func add(item: T) {
```

Comment:
> but this class should have a subscript

Should it? As mentioned in the comments, ordering is not guaranteed since we can reuse holes in the array. I'm not sure how useful direct indexing would be. Also, the implementation would be a bit complicated since the index mappings would change as items are deallocated.

> This should probably be "append" though to be consistent with arrays.

Also not sure I agree for the same reason -- ordering isn't preserved. "Append" generally means "add to the end".